tests/integration/new_relations/test_relations_coherence.py

```diff
@@ -0,0 +1,127 @@
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_relations(ops_test: OpsTest, charm):
+    """Test that check relation data."""
+    async with ops_test.fast_forward():
+        # Relate with client-app, wait-for-ready, check relation data, remove-relation
+        await ops_test.model.add_relation(
+            f"{APPLICATION_APP_NAME}:{FIRST_DATABASE_RELATION_NAME}", DATABASE_APP_NAME
+        )
+        await ops_test.model.wait_for_idle(apps=APP_NAMES, status="active")
+
+        primary_connection_string = await build_connection_string(
+            ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME
+        )
+
+        with psycopg2.connect(primary_connection_string) as connection:
+            connection.autocommit = True
+            with connection.cursor() as cursor:
+                # Check that it's possible to write and read data from the database that
+                # was created for the application.
+                cursor.execute("DROP TABLE IF EXISTS test;")
+                cursor.execute("CREATE TABLE test(data TEXT);")
+                cursor.execute("INSERT INTO test(data) VALUES('some data');")
+                cursor.execute("SELECT data FROM test;")
+                data = cursor.fetchone()
+                assert data[0] == "some data"
+        connection.close()
+
+        await ops_test.model.applications[DATABASE_APP_NAME].remove_relation(
+            f"{DATABASE_APP_NAME}:database",
+            f"{APPLICATION_APP_NAME}:{FIRST_DATABASE_RELATION_NAME}",
+        )
+        await ops_test.model.wait_for_idle(apps=[DATABASE_APP_NAME], status="active", timeout=1000)
+
+        # Relate with client-app 2nd time, wait-for-ready, check relation data
+        await ops_test.model.add_relation(
+            f"{APPLICATION_APP_NAME}:{FIRST_DATABASE_RELATION_NAME}", DATABASE_APP_NAME
+        )
+        await ops_test.model.wait_for_idle(apps=APP_NAMES, status="active")
+
+        primary_connection_string = await build_connection_string(
+            ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME
+        )
+        with psycopg2.connect(primary_connection_string) as connection:
+            connection.autocommit = True
+            with connection.cursor() as cursor:
+                cursor.execute("SELECT data FROM test;")
+                data = cursor.fetchone()
+                assert data[0] == "some data"
+        connection.close()
+
+        await ops_test.model.applications[DATABASE_APP_NAME].remove_relation(
+            f"{DATABASE_APP_NAME}:database",
+            f"{APPLICATION_APP_NAME}:{FIRST_DATABASE_RELATION_NAME}",
+        )
+        await ops_test.model.wait_for_idle(apps=[DATABASE_APP_NAME], status="active", timeout=1000)
+
+        # Repeat re-relation using all available relation options (e.g. legacy_roles)
+        await ops_test.model.applications[APPLICATION_APP_NAME].set_config({
+            "legacy_roles": "true",
+        })
```

Comment:
This won't work. The `legacy_roles` is there to check that the postgresql charm will block if those are requested and there is no way to satisfy that interface.

Reply:
Alternative roles that can be used with the modern relation are, for example: `admin` (custom role created by the charm) and PostgreSQL predefined roles like `CREATEDB`, `CREATEROLE` and `SUPERUSER`.

Reply:
It was requested `extra-user-roles` [here](https://github.com/canonical/postgresql-operator/issues/449) to make sure switching from user->admin->user->admin works as expected. Admin has admin level and User has no admin level => test to make sure no garbage left in DB/databags.